familyApp/app/components/Finance.js

```diff
@@ -31,92 +19,159 @@ export default class Finance extends React.Component{
     super();
     //all the data save before to can show in the bar
     this.state={
-      //defult thing when change from data base change here 🙂 <3
-      //female (Mother) / male (Father)/ child (Children)
-      userType:'female',
-      //from 100%
-      userProgress:'100',
-      //for how many child in family
-      userRanking:'2',
-      //the money still
-      restMoney:'1500',
-      //
       tableHead:  ['Name', 'Cost'],
-      tableName: ['Water', 'Electricity', 'Shortage', 'Family Event'],
-      tableCost:  [[12],     [20],            [40],         [50]    ],
-      tableTotal:['Total',0]
+      tableName: [['Water'], ['Electricity'], ['Shortage'], ['Family Event'],['water of the month']],
+      tableCost:  [[12],     [30],            [40],         [150]    ,[999.99]],
+      tableTotal:['Total',0],
+      //for show Dialog Add
+      addDialogVisible: false,
+      addName:'',
+      addEditCost:'',
+      editDialogVisible: false,
+      editCost:'',
+      deleteDialogVisible:false,
+
     };
     //auto call function when render this scren
-    this.calculateTotal();
+    this.calculateTotalMoney();
   }
-  calculateTotal(){
+  calculateTotalMoney(){
     var total=0;
     for (var i = 0; i < this.state.tableCost.length; i++) {
       total+=this.state.tableCost[i][0];
     }
     //cant use set state so we use this .state
     this.state.tableTotal[1]=total;
-    //this.setState({tableTotal : ['Total',total]});
-    //alert('you cal total: '+this.state.tableTotal[1]);
-  }
-  fectch1(){
-    //return axios.get('http://192.168.1.82:3000')
-    return fetch('http://192.168.1.82:3000')
-      .then((response) => response.json())
-        .then((responseJson) => {
-          console.log("server done:",JSON.stringify(responseJson) )
-           alert(JSON.stringify(responseJson));
-        })
-      .catch(function (error) {
-       console.log(error);
-      });
+    // this.setState({tableTotal : ['Total',total]});
   }
+  handleCancelAdd(){
+    this.setState({ addDialogVisible: false });
+  };
+  handleAdd(){
+    alert('Done Add: '+this.state.addName +' with cost: ' + this.state.addEditCost);
+    this.setState({ addDialogVisible: false });
+  };
   addToFinance(){
-    alert('Add To Finance');
+    //alert('Add To Finance');
+    this.setState({ addDialogVisible: true });
+  };
+  onAddName(name){
+    if(name.length>0){
+      name=name[0].toUpperCase()+name.slice(1,name.length)
+    }
+    this.setState({addName: name})
+  }
+  onAddEditCost(value) {
```

Comment:
This looks like it was a pain to code. A general word of advice is that, if you think the problem you have is something that a million other people have also had to face, there is probably some package that already exists to solve your problem. There's a saying in English that goes, "You don't need to re-invent the wheel."

For validating your inputs: https://medium.freecodecamp.org/how-to-use-reacts-controlled-inputs-for-instant-form-field-validation-b1c7b033527e
For handling number formats in React: https://www.npmjs.com/package/react-number-format

All of this took me a few minutes to find on Google. Do some googling next time, and it can save you hours of coding. Plus, these pre-existing packages have probably already been tested by a ton of people, so they're probably pretty reliable.

Also, as another general note, it is usually much better to take any money-related input from the user as a number, rather than as a string. Strings are very hard to control. Take the input as a number and then, if necessary, display it as a formatted string using the packages above.

Reply:
yes thank you but some time I take a small code from internet and then I developed 
so I save half of my time and I learn and understand the code better in the another half of my time

and abut the number I try to find way but I dont so before I sent it I convert it to number by JSON.parse()
and in my case I check I ignore any non logical input as a number and it is work
thank you for you follow and advice

if you have any other advice I am so wait for it (I really like to learn from how is more experience than me ♥)
and thank you to be part of RBK

Reply:
but I will start google more things i need and spent quarter of my time to find the code do what i need in google and only quarter of my time to understand the code to make my ogresses faster ☺